com.hazelcast.jet.core.OperationLossTest.when_connectionDroppedWithoutMemberLeaving_then_jobRestarts

_master_

Failed on PR builder for nightly tests: http://jenkins.hazelcast.com/job/jet-oss-pr-builder-nightly-tests/10/testReport/com.hazelcast.jet.core/OperationLossTest/when_connectionDroppedWithoutMemberLeaving_then_jobRestarts/

Stacktrace:
```
java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at java.util.concurrent.CompletableFuture.reportJoin(CompletableFuture.java:375)
	at java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
	at com.hazelcast.jet.Job.join(Job.java:155)
	at com.hazelcast.jet.core.OperationLossTest.when_connectionDroppedWithoutMemberLeaving_then_jobRestarts(OperationLossTest.java:173)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:373)
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346)
	... 1 more
Caused by: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148)
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366)
	... 3 more
```

Standard output:
```
2020-09-21 13:28:13,945 [ INFO] [when_connectionDroppedWithoutMemberLeaving_then_jobRestarts] [c.h.t.m.MockNetworkingService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=false}
connection closed
2020-09-21 13:28:13,969 [ WARN] [hz.infallible_austin.jet.cooperative.thread-28] [c.h.j.i.e.TaskletExecutionService]: [127.0.0.1]:5702 [jet] [4.3-SNAPSHOT] Exception in ReceiverTasklet
java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) [?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) [classes/:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
2020-09-21 13:28:13,973 [DEBUG] [hz.infallible_austin.jet.blocking.thread-0] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.3-SNAPSHOT] Execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:607) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) [?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) [?:1.8.0_251]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:59) [classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:461) [classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$BlockingWorker.run(TaskletExecutionService.java:309) [classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_251]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_251]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_251]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_251]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
Caused by: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:373) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) ~[classes/:?]
	... 1 more
Caused by: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) ~[classes/:?]
	... 1 more
2020-09-21 13:28:13,975 [ INFO] [hz.infallible_austin.jet.blocking.thread-0] [c.h.t.m.MockNetworkingService]: [127.0.0.1]:5702 [jet] [4.3-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5701, alive=true}
2020-09-21 13:28:14,028 [ WARN] [hz.romantic_austin.jet.cooperative.thread-34] [c.h.j.i.e.TaskletExecutionService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Exception in ReceiverTasklet
java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5702
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) [?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) [classes/:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
2020-09-21 13:28:14,029 [DEBUG] [hz.romantic_austin.jet.cooperative.thread-32] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001 completed with failure
java.util.concurrent.CompletionException: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5702
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:607) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488) ~[?:1.8.0_251]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:59) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:461) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:392) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:378) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) [?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) [classes/:?]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_251]
Caused by: com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5702
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:373) ~[classes/:?]
	... 3 more
Caused by: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5702
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366) ~[classes/:?]
	... 3 more
2020-09-21 13:28:14,032 [ INFO] [hz.romantic_austin.async.thread-4] [c.h.t.m.MockNetworkingService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Created connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=true}
2020-09-21 13:28:14,076 [DEBUG] [ForkJoinPool.commonPool-worker-66] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001 has failures: [[127.0.0.1]:5702=com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701, [127.0.0.1]:5701=com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5702]
2020-09-21 13:28:14,076 [DEBUG] [hz.romantic_austin.cached.thread-9] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Sending CompleteExecutionOperation for job '0500-7539-2509-0001', execution 0500-7539-250a-0001
2020-09-21 13:28:14,078 [DEBUG] [hz.romantic_austin.cached.thread-9] [c.h.j.i.o.CompleteExecutionOperation]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Completing execution 0500-7539-250a-0001 from caller [127.0.0.1]:5701, error=com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
2020-09-21 13:28:14,078 [DEBUG] [hz.romantic_austin.cached.thread-9] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Completed execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001
2020-09-21 13:28:14,078 [DEBUG] [hz.infallible_austin.generic-operation.thread-20] [c.h.j.i.o.CompleteExecutionOperation]: [127.0.0.1]:5702 [jet] [4.3-SNAPSHOT] Completing execution 0500-7539-250a-0001 from caller [127.0.0.1]:5701, error=com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
2020-09-21 13:28:14,078 [DEBUG] [hz.infallible_austin.generic-operation.thread-20] [c.h.j.i.JobExecutionService]: [127.0.0.1]:5702 [jet] [4.3-SNAPSHOT] Completed execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001
2020-09-21 13:28:14,078 [ERROR] [hz.romantic_austin.cached.thread-9] [c.h.j.i.MasterJobContext]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] Execution of job '0500-7539-2509-0001', execution 0500-7539-250a-0001 failed
	Start time: 2020-09-21T13:28:13.743
	Duration: 335 ms
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
com.hazelcast.jet.JetException: Exception in ReceiverTasklet: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:373) ~[classes/:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:891) ~[?:1.8.0_251]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:346) ~[classes/:?]
	at java.lang.Thread.run(Thread.java:748) ~[?:1.8.0_251]
Caused by: java.lang.RuntimeException: The member was reconnected: [127.0.0.1]:5701
	at com.hazelcast.jet.impl.execution.ReceiverTasklet.call(ReceiverTasklet.java:148) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:366) ~[classes/:?]
	... 3 more
2020-09-21 13:28:14,079 [DEBUG] [hz.romantic_austin.cached.thread-9] [c.h.j.i.JobCoordinationService]: [127.0.0.1]:5701 [jet] [4.3-SNAPSHOT] job '0500-7539-2509-0001', execution 0500-7539-250a-0001 is completed
BuildInfo right after when_connectionDroppedWithoutMemberLeaving_then_jobRestarts(com.hazelcast.jet.core.OperationLossTest): BuildInfo{version='4.0.3-SNAPSHOT', build='20200914', buildNumber=20200914, revision=d717127, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='4.3-SNAPSHOT', build='20200921', revision='58e4a39'}}
Hiccups measured while running test 'when_connectionDroppedWithoutMemberLeaving_then_jobRestarts(com.hazelcast.jet.core.OperationLossTest):'
13:28:10, accumulated pauses: 245 ms, max pause: 200 ms, pauses over 1000 ms: 0
13:28:15, accumulated pauses: 0 ms, max pause: 0 ms, pauses over 1000 ms: 0
```

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/AbstractJobProxy.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/AbstractJobProxy.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.jet.impl;
 
+import com.hazelcast.core.HazelcastInstanceNotActiveException;
 import com.hazelcast.core.LocalMemberResetException;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.internal.serialization.Data;
@@ -193,18 +194,26 @@ public abstract class AbstractJobProxy<T> implements Job {
 
     protected abstract JobConfig doGetJobConfig();
 
+    /**
+     * Get the current master UUID.
+     *
+     * @throws IllegalStateException if the master isn't known
+     */
+    @Nonnull
     protected abstract UUID masterUuid();
 
     protected abstract SerializationService serializationService();
 
     protected abstract LoggingService loggingService();
 
+    protected abstract boolean isRunning();
+
     protected T container() {
         return container;
     }
 
     private void doSubmitJob(Object jobDefinition, JobConfig config) {
-        CompletableFuture<Void> submitFuture = new CompletableFuture<>();
+        NonCompletableFuture submitFuture = new NonCompletableFuture();
         SubmitJobCallback callback = new SubmitJobCallback(submitFuture, jobDefinition, config);
         invokeSubmitJob(serializationService().toData(jobDefinition), config).whenCompleteAsync(callback);
         submitFuture.join();
@@ -213,97 +222,100 @@ public abstract class AbstractJobProxy<T> implements Job {
     private boolean isRestartable(Throwable t) {
         return t instanceof MemberLeftException
                 || t instanceof TargetDisconnectedException
-                || t instanceof TargetNotMemberException;
+                || t instanceof TargetNotMemberException
+                || t instanceof HazelcastInstanceNotActiveException && isRunning();
     }
 
     private void doInvokeJoinJob() {
         invokeJoinJob().whenCompleteAsync(joinJobCallback);
     }
 
-    private class SubmitJobCallback implements BiConsumer<Void, Throwable> {
-        private final CompletableFuture<Void> future;
-        private final Object jobDefinition;
-        private final JobConfig config;
+    private abstract class CallbackBase implements BiConsumer<Void, Throwable> {
+        private final NonCompletableFuture future;
 
-        SubmitJobCallback(CompletableFuture<Void> future, Object jobDefinition, JobConfig config) {
+        protected CallbackBase(NonCompletableFuture future) {
             this.future = future;
-            this.jobDefinition = jobDefinition;
-            this.config = config;
         }
 
         @Override
-        public void accept(Void aVoid, Throwable t) {
+        public final void accept(Void aVoid, Throwable t) {
             if (t != null) {
                 Throwable ex = peel(t);
                 if (ex instanceof LocalMemberResetException) {
-                    String msg = "Submission of job " + idAndName() + " failed because the cluster is performing " +
-                            "split-brain merge";
-                    logger.warning(msg, ex);
-                    future.completeExceptionally(new CancellationException(msg));
-                } else if (!isRestartable(ex)) {
-                    future.completeExceptionally(ex);
-                } else {
-                    try {
-                        resubmitJob(t);
-                    } catch (Exception e) {
-                        future.completeExceptionally(peel(e));
-                    }
-                }
-            } else {
-                future.complete(null);
-            }
-        }
-
-        private void resubmitJob(Throwable t) {
-            if (masterUuid() != null) {
-                logger.fine("Resubmitting job " + idAndName() + " after " + t.getClass().getSimpleName());
-                invokeSubmitJob(serializationService().toData(jobDefinition), config).whenCompleteAsync(this);
-                return;
-            }
-            // job data will be cleaned up eventually by coordinator
-            String msg = "Job " + idAndName() + " failed because the cluster is performing "
-                    + " split-brain merge and coordinator is not known";
-            logger.warning(msg, t);
-            future.completeExceptionally(new CancellationException(msg));
-        }
-    }
-
-    private class JoinJobCallback implements BiConsumer<Void, Throwable> {
-
-        @Override
-        public void accept(Void aVoid, Throwable t) {
-            if (t != null) {
-                Throwable ex = peel(t);
-                if (ex instanceof LocalMemberResetException) {
-                    String msg = "Job " + idAndName() + " failed because the cluster is performing a split-brain merge";
+                    String msg = operationName() + " failed for job " + idAndName()
+                            + " because the cluster is performing split-brain merge";
                     logger.warning(msg, ex);
                     future.internalCompleteExceptionally(new CancellationException(msg));
                 } else if (!isRestartable(ex)) {
                     future.internalCompleteExceptionally(ex);
                 } else {
                     try {
-                        rejoinJob(t);
+                        retryAction(t);
                     } catch (Exception e) {
                         future.internalCompleteExceptionally(peel(e));
                     }
                 }
             } else {
-                // job completed successfully
                 future.internalComplete();
             }
         }
 
-        private void rejoinJob(Throwable t) {
-            if (masterUuid() != null) {
-                logger.fine("Rejoining to job " + idAndName() + " after " + t.getClass().getSimpleName());
-                doInvokeJoinJob();
+        private void retryAction(Throwable t) {
+            try {
+                // calling for the side-effect of throwing ISE if master not known
+                masterUuid();
+            } catch (IllegalStateException e) {
+                // job data will be cleaned up eventually by the coordinator
+                String msg = operationName() + " failed for job " + idAndName() + " because the cluster " +
+                        "is performing  split-brain merge and the coordinator is not known";
+                logger.warning(msg, t);
+                future.internalCompleteExceptionally(new CancellationException(msg));
                 return;
             }
-            // job data will be cleaned up eventually by coordinator
-            String msg = "Job " + idAndName() + " failed because the cluster is performing "
-                    + " split-brain merge and coordinator is not known";
-            logger.warning(msg, t);
-            future.internalCompleteExceptionally(new CancellationException(msg));
+            retryActionInt(t);
+        }
+
+        protected abstract void retryActionInt(Throwable t);
+        protected abstract String operationName();
+    }
+
+    private class SubmitJobCallback extends CallbackBase {
+        private final Object jobDefinition;
+        private final JobConfig config;
+
+        SubmitJobCallback(NonCompletableFuture future, Object jobDefinition, JobConfig config) {
+            super(future);
+            this.jobDefinition = jobDefinition;
+            this.config = config;
+        }
+
+        @Override
+        protected void retryActionInt(Throwable t) {
+            logger.fine("Resubmitting job " + idAndName() + " after " + t.getClass().getSimpleName());
+            invokeSubmitJob(serializationService().toData(jobDefinition), config).whenCompleteAsync(this);
+        }
+
+        @Override
+        protected String operationName() {
+            return "Submit";
+        }
+    }
+
+    private class JoinJobCallback extends CallbackBase {
+
+        JoinJobCallback() {
+            super(AbstractJobProxy.this.future);
+        }
+
+        @Override
+        protected void retryActionInt(Throwable t) {
+            logger.fine("Rejoining to job " + idAndName() + " after " + t.getClass().getSimpleName(), t);
+            doInvokeJoinJob();
+        }
+
+        @Override
+        protected String operationName() {
+            return "Join";
         }
     }
 }

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ClientJobProxy.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/ClientJobProxy.java
@@ -18,6 +18,7 @@ package com.hazelcast.jet.impl;
 
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.spi.impl.ClientInvocation;
+import com.hazelcast.cluster.Member;
 import com.hazelcast.internal.serialization.Data;
 import com.hazelcast.internal.serialization.SerializationService;
 import com.hazelcast.jet.Job;
@@ -170,9 +171,13 @@ public class ClientJobProxy extends AbstractJobProxy<JetClientInstanceImpl> {
         });
     }
 
-    @Override
+    @Nonnull @Override
     protected UUID masterUuid() {
-        return container().getHazelcastClient().getClientClusterService().getMasterMember().getUuid();
+        Member masterMember = container().getHazelcastClient().getClientClusterService().getMasterMember();
+        if (masterMember == null) {
+            throw new IllegalStateException("Master isn't known");
+        }
+        return masterMember.getUuid();
     }
 
     @Override
@@ -183,6 +188,11 @@ public class ClientJobProxy extends AbstractJobProxy<JetClientInstanceImpl> {
     @Override
     protected LoggingService loggingService() {
         return container().getHazelcastClient().getLoggingService();
+    }
+
+    @Override
+    protected boolean isRunning() {
+        return container().getHazelcastClient().getLifecycleService().isRunning();
     }
 
     private ClientInvocation invocation(ClientMessage request, UUID invocationUuid) {

--- a/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobProxy.java
+++ b/hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobProxy.java
@@ -154,7 +154,7 @@ public class JobProxy extends AbstractJobProxy<NodeEngineImpl> {
         }
     }
 
-    @Override
+    @Nonnull @Override
     protected UUID masterUuid() {
         Collection<Member> members = container().getClusterService().getMembers();
         if (members.isEmpty()) {
@@ -171,6 +171,11 @@ public class JobProxy extends AbstractJobProxy<NodeEngineImpl> {
     @Override
     protected LoggingService loggingService() {
         return container().getLoggingService();
+    }
+
+    @Override
+    protected boolean isRunning() {
+        return container().isRunning();
     }
 
     private <T> CompletableFuture<T> invokeOp(Operation op) {

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/SimpleTestInClusterSupport.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/SimpleTestInClusterSupport.java
@@ -20,6 +20,8 @@ import com.hazelcast.core.DistributedObject;
 import com.hazelcast.jet.config.JetClientConfig;
 import com.hazelcast.jet.config.JetConfig;
 import com.hazelcast.jet.core.JetTestSupport;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
 import com.hazelcast.test.HazelcastSerialClassRunner;
 import org.junit.After;
 import org.junit.AfterClass;
@@ -27,7 +29,13 @@ import org.junit.runner.RunWith;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.List;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static com.hazelcast.jet.Util.idToString;
+import static java.util.stream.Collectors.joining;
 
 /**
  * Base class for tests that share the cluster for all jobs. The subclass must
@@ -35,6 +43,8 @@ import java.util.concurrent.TimeUnit;
  */
 @RunWith(HazelcastSerialClassRunner.class)
 public abstract class SimpleTestInClusterSupport extends JetTestSupport {
+
+    private static final ILogger SUPPORT_LOGGER = Logger.getLogger(SimpleTestInClusterSupport.class);
 
     private static JetTestInstanceFactory factory;
     private static JetConfig jetConfig;
@@ -74,10 +84,18 @@ public abstract class SimpleTestInClusterSupport extends JetTestSupport {
             return;
         }
         // after each test ditch all jobs and objects
-        for (Job job : instances[0].getJobs()) {
+        List<Job> jobs = instances[0].getJobs();
+        SUPPORT_LOGGER.info("Ditching " + jobs.size() + " jobs in SimpleTestInClusterSupport.@After: " +
+                jobs.stream().map(j -> idToString(j.getId())).collect(joining(", ", "[", "]")));
+        for (Job job : jobs) {
             ditchJob(job, instances());
         }
-        for (DistributedObject o : instances()[0].getHazelcastInstance().getDistributedObjects()) {
+        Collection<DistributedObject> objects = instances()[0].getHazelcastInstance().getDistributedObjects();
+        SUPPORT_LOGGER.info("Destroying " + objects.size()
+                + " distributed objects in SimpleTestInClusterSupport.@After: "
+                + objects.stream().map(o -> o.getServiceName() + "/" + o.getName())
+                         .collect(Collectors.joining(", ", "[", "]")));
+        for (DistributedObject o : objects) {
             o.destroy();
         }
     }
@@ -85,6 +103,7 @@ public abstract class SimpleTestInClusterSupport extends JetTestSupport {
     @AfterClass
     public static void supportAfterClass() throws Exception {
         if (factory != null) {
+            SUPPORT_LOGGER.info("Terminating instance factory in SimpleTestInClusterSupport.@AfterClass");
             spawn(() -> factory.terminateAll())
                     .get(1, TimeUnit.MINUTES);
         }

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JetTestSupport.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/JetTestSupport.java
@@ -68,12 +68,15 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
     @ClassRule
     public static Timeout globalTimeout = Timeout.seconds(15 * 60);
 
+    private static final ILogger SUPPORT_LOGGER = Logger.getLogger(JetTestSupport.class);
+
     protected ILogger logger = Logger.getLogger(getClass());
     private JetTestInstanceFactory instanceFactory;
 
     @After
     public void shutdownFactory() throws Exception {
         if (instanceFactory != null) {
+            SUPPORT_LOGGER.info("Terminating instanceFactory in JetTestSupport.@After");
             spawn(() -> instanceFactory.terminateAll())
                     .get(1, TimeUnit.MINUTES);
         }
@@ -236,7 +239,7 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
             try {
                 r.runEx();
             } catch (Throwable e) {
-                logger.warning("Spawned Runnable failed", e);
+                SUPPORT_LOGGER.warning("Spawned Runnable failed", e);
             }
         });
     }
@@ -255,7 +258,7 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
             assertTrue("stats are 0", allowEmptySnapshot || record.snapshotStats().numBytes() > 0);
             snapshotId[0] = record.snapshotId();
         }, timeoutSeconds);
-        logger.info("First snapshot found (id=" + snapshotId[0] + ")");
+        SUPPORT_LOGGER.info("First snapshot found (id=" + snapshotId[0] + ")");
     }
 
     public void waitForNextSnapshot(JobRepository jr, long jobId, int timeoutSeconds, boolean allowEmptySnapshot) {
@@ -271,7 +274,7 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
                     snapshotId[0] > originalSnapshotId);
             assertTrue("stats are 0", allowEmptySnapshot || record.snapshotStats().numBytes() > 0);
         }, timeoutSeconds);
-        logger.info("Next snapshot found after " + NANOSECONDS.toMillis(System.nanoTime() - start) + " ms (id="
+        SUPPORT_LOGGER.info("Next snapshot found after " + NANOSECONDS.toMillis(System.nanoTime() - start) + " ms (id="
                 + snapshotId[0] + ", previous id=" + originalSnapshotId + ")");
     }
 
@@ -306,7 +309,7 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
                     return;
                 }
             } catch (Exception e) {
-                logger.warning("Failure to read job status: " + e, e);
+                SUPPORT_LOGGER.warning("Failure to read job status: " + e, e);
             }
 
             Exception cancellationFailure;
@@ -324,8 +327,8 @@ public abstract class JetTestSupport extends HazelcastTestSupport {
             }
 
             sleepMillis(500);
-            logger.warning("Failed to cancel the job and it is " + status + ", retrying. Failure: " + cancellationFailure,
-                    cancellationFailure);
+            SUPPORT_LOGGER.warning("Failed to cancel the job and it is " + status + ", retrying. Failure: "
+                            + cancellationFailure, cancellationFailure);
         }
         // if we got here, 10 attempts to cancel the job have failed. Cluster is in bad shape probably, shut it down
         try {

--- a/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
+++ b/hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/TopologyChangeTest.java
@@ -96,7 +96,7 @@ public class TopologyChangeTest extends JetTestSupport {
     private JetInstance[] instances;
     private JetConfig config;
 
-    @Parameterized.Parameters(name = "liteMemberFlags({index}")
+    @Parameterized.Parameters(name = "liteMemberFlags({index})")
     public static Collection<boolean[]> parameters() {
         return Arrays.asList(new boolean[][]{
                 {false, false, false},

